core/runtime/src/url.rs

```diff
@@ -0,0 +1,237 @@
+//! Boa's implementation of JavaScript's `URL` Web API class.
+//!
+//! The `URL` class can be instantiated from any global object.
+//! This relies on the `url` feature.
+//!
+//! More information:
+//!  - [MDN documentation][mdn]
+//!  - [WHATWG `URL` specification][spec]
+//!
+//! [spec]: https://url.spec.whatwg.org/
+//! [mdn]: https://developer.mozilla.org/en-US/docs/Web/API/URL
+#![cfg(feature = "url")]
+
+#[cfg(test)]
+mod tests;
+
+use boa_engine::value::Convert;
+use boa_engine::{
+    js_error, js_string, Context, Finalize, JsData, JsResult, JsString, JsValue, Trace,
+};
+use boa_interop::{js_class, IntoJsFunctionCopied, JsClass};
+use std::fmt::Display;
+
+/// The `URL` class represents a (properly parsed) Uniform Resource Locator.
+#[derive(Debug, Clone, JsData, Trace, Finalize)]
+pub struct Url(#[unsafe_ignore_trace] url::Url);
+
+impl Url {
+    /// Register the `URL` class into the realm.
+    ///
+    /// # Errors
+    /// This will error if the context or realm cannot register the class.
+    pub fn register(context: &mut Context) -> JsResult<()> {
+        context.register_global_class::<Self>()?;
+        Ok(())
+    }
+
+    /// Create a new `URL` object. Meant to be called from the JavaScript constructor.
+    ///
+    /// # Errors
+    /// Any errors that might occur during URL parsing.
+    fn js_new(Convert(ref url): Convert<String>, base: &Option<Convert<String>>) -> JsResult<Self> {
+        if let Some(Convert(base)) = base {
+            let base_url = url::Url::parse(base)
+                .map_err(|e| js_error!(TypeError: "Failed to parse base URL: {}", e))?;
+            if base_url.cannot_be_a_base() {
+                return Err(js_error!(TypeError: "Base URL {} cannot be a base", base));
+            }
+
+            let url = base_url
+                .join(url)
+                .map_err(|e| js_error!(TypeError: "Failed to parse URL: {}", e))?;
+            Ok(Self(url))
+        } else {
+            let url = url::Url::parse(url)
+                .map_err(|e| js_error!(TypeError: "Failed to parse URL: {}", e))?;
+            Ok(Self(url))
+        }
+    }
+}
+
+impl Display for Url {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl From<url::Url> for Url {
+    fn from(url: url::Url) -> Self {
+        Self(url)
+    }
+}
+
+impl From<Url> for url::Url {
+    fn from(url: Url) -> url::Url {
+        // Cannot avoid cloning here, unfortunately, as we would need to replace
+        // the internal URL with something else.
+        url.0.clone()
```

Comment:
For this we can use the `#[boa_gc(unsafe_no_drop)]` on the Url which will not generate a Drop `impl` for `Url` and the value can be moved out.